### PR TITLE
JSTL - studentHome

### PR DESCRIPTION
**Pull request by yj-soh**
studentHome.jsp with JSTL.
HTML tests not modified yet.

### CODE REVIEW

File: src/main/java/teammates/ui/controller/StudentHomePageAction.java [hunk 1]
```diff
@@ -25,32 +26,36 @@ public ActionResult execute() throws EntityDoesNotExistException {
...
+            courses = logic.getCourseDetailsListForStudent(account.googleId);
+            sessionSubmissionStatusMap = generateFeedbackSessionSubmissionStatusMap(courses, account.googleId);
             
-            statusToAdmin = "studentHome Page Load<br>" + "Total courses: " + data.courses.size();
+            CourseDetailsBundle.sortDetailedCourses(courses);
             
-            boolean isDataConsistent = checkEventualConsistency(recentlyJoinedCourseId);
+            statusToAdmin = "studentHome Page Load<br>" + "Total courses: " + courses.size();
+            
+            boolean isDataConsistent = checkEventualConsistency(recentlyJoinedCourseId, courses);
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/StudentHomePageAction.java]: The method too should be named `isDataConsistent(courseId,courses)` ? May be a better name is `isCourseIncluded(courseId,courses)` because that's what the method checks.

File: src/main/java/teammates/ui/controller/StudentHomePageAction.java [hunk 2]
```diff
@@ -103,16 +108,18 @@ private boolean checkEventualConsistency(String recentlyJoinedCourseId) {
     }
 
     private void showEventualConsistencyMessage(String recentlyJoinedCourseId) {
-        String errorMessage = String.format(Const.StatusMessages.EVENTUAL_CONSISTENCY_MESSAGE_STUDENT, recentlyJoinedCourseId);
+        String errorMessage = String.format(Const.StatusMessages.EVENTUAL_CONSISTENCY_MESSAGE_STUDENT,
+                                            recentlyJoinedCourseId);
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/StudentHomePageAction.java]: Resist doing unrelated changes like this one. especially layout tweaks.

File: src/test/java/teammates/test/cases/ui/StudentHomePageActionTest.java [hunk 3]
```diff
@@ -160,7 +160,7 @@ public void testExecuteAndPostProcess() throws Exception{
         a = getAction(submissionParams);
         r = getShowPageResult(a);
         data = (StudentHomePageData) r.data;
-        assertEquals(1, data.courses.size());
+        assertEquals(1, data.getCourseTables().size());
         
         
         // Delete additional sessions that were created
```
Comment (damithc) [inline src/test/java/teammates/test/cases/ui/StudentHomePageActionTest.java]: I was expecting to see some new tests. Is the coverage good? I think we should add a `*PageDataTest` class.

**Comment by yj-soh** [comment]
Added the test.
99%+ coverage on `StudentHomePageData`; only misses a couple of branches in that block of HTML.
That HTML block part is untested, but I mentioned it in a TODO.

File: src/main/java/teammates/ui/controller/StudentHomePageData.java [hunk 4]
```diff
@@ -42,15 +96,15 @@ public String getStudentStatusForSession(FeedbackSessionAttributes session){
     }
     
     /**
+     * @param sessionSubmissionStatusMap 
      * @param submissionStatus Submission status of a student for a particular feedback session. 
      * 
      * @return The hover message to explain evaluation submission status.
      */
-    public String getStudentHoverMessageForSession(FeedbackSessionAttributes session){
+    private String getStudentHoverMessageForSession(FeedbackSessionAttributes session, boolean hasSubmitted){
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/StudentHomePageData.java]: The parameters to this method do not match the javadoc comments

File: src/main/java/teammates/ui/controller/StudentHomePageData.java [hunk 5]
```diff
@@ -94,11 +148,10 @@ public String getStudentFeedbackResultsLink(String courseId, String feedbackSess
     
 
     /**
+     * @param sessionSubmissionStatusMap 
      * @return The list of available actions for a specific feedback session.
      */
-    public String getStudentFeedbackSessionActions(FeedbackSessionAttributes fs, int idx) {
-        String keyOfMap = fs.courseId+"%"+fs.feedbackSessionName;
-        boolean hasSubmitted = sessionSubmissionStatusMap.get(keyOfMap).booleanValue();
+    private String getStudentFeedbackSessionActions(FeedbackSessionAttributes fs, int idx, boolean hasSubmitted) {
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/StudentHomePageData.java]: Same for this

**Comment by yj-soh** [comment]
Passes all related tests and ready for review, for anyone interested.

File: src/test/java/teammates/test/cases/ui/pagedata/StudentHomePageDataTest.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                            "Closed");
+    }
+    
+    private void testFeedbackSession(Map<String, Object> row, FeedbackSessionAttributes session,
+            String expectedTooltip, String expectedStatus) {
+        assertEquals(session.feedbackSessionName, row.get("name"));
+        assertEquals(TimeHelper.formatTime(session.endTime), row.get("endTime"));
+        assertEquals(expectedTooltip, row.get("tooltip"));
+        assertEquals(expectedStatus, row.get("status"));
+        //TODO: verify actions are correct after they are not merely a block of HTML
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/pagedata/StudentHomePageDataTest.java]: Now that #3823 is done, you can test this now?

File: src/test/java/teammates/test/cases/ui/pagedata/StudentHomePageDataTest.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+        FeedbackSessionAttributes session = new FeedbackSessionAttributes();
+        session.feedbackSessionName = name;
+        session.startTime = TimeHelper.getHoursOffsetToCurrentTime(offsetStart);
+        session.endTime = TimeHelper.getHoursOffsetToCurrentTime(offsetEnd);
+        session.resultsVisibleFromTime = TimeHelper.getHoursOffsetToCurrentTime(offsetPublish);
+        session.sessionVisibleFromTime = TimeHelper.getHoursOffsetToCurrentTime(-1);
+        return session;
+    }
+
+    private CourseDetailsBundle createCourseBundle(CourseAttributes course, FeedbackSessionAttributes... sesssions) {
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/pagedata/StudentHomePageDataTest.java]: You spelled sessions wrongly

**Comment by junhaoyap** [comment]
@kanghj wanna finish reviewing this? :D
